### PR TITLE
Use username for comparison only when impersonation is enabled

### PR DESCRIPTION
**Pull request by NikhilCollooru**
When impersonation is NOT enabled, we still want the MetastoreContext to have correct username and not a dummy value. So this PR moves the impersonation logic into `KeyAndContext` . The `equals()` and `hashCode()` methods of KeyAndContext will use username ONLY when impersonation is enabled. 

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java [hunk 1]
```diff
@@ -938,10 +946,9 @@ public String toString()
         }
     }
 
-    private <T> KeyAndContext<T> getCachingKey(MetastoreContext context, T key)
+    private <T> KeyAndContext<T> getCachingKey(MetastoreContext metastoreContext, T key)
     {
-        MetastoreContext metastoreContext = metastoreImpersonationEnabled ? context : new MetastoreContext(NO_IMPERSONATION_USER, context.getQueryId(), context.getClientInfo(), context.getSource());
-        return new KeyAndContext<>(metastoreContext, key);
+        return new KeyAndContext<>(metastoreContext, key, metastoreImpersonationEnabled);
```
Comment (shixuan-fan) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:951]: How about we make `impersonationEnabled` part of `MetastoreContext`? This way we don't need an extra parameter here, and `KeyAndContext` can use the context information to determine its `equals` and `hashCode`.

Comment (NikhilCollooru) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:951]: yeah I agree. Since `impersonation` is related to Metastore, it is better add it to `MetastoreContext`.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java [hunk 2]
```diff
@@ -940,7 +946,7 @@ public String toString()
 
     private <T> KeyAndContext<T> getCachingKey(MetastoreContext context, T key)
     {
-        MetastoreContext metastoreContext = metastoreImpersonationEnabled ? context : new MetastoreContext(NO_IMPERSONATION_USER, context.getQueryId(), context.getClientInfo(), context.getSource());
+        MetastoreContext metastoreContext = metastoreImpersonationEnabled ? new MetastoreContext(context.getUsername(), context.getQueryId(), context.getClientInfo(), context.getSource(), true) : context;
```
Comment (shixuan-fan) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:949]: Mutating the context looks a bit weird , but I could see it might be necessary we don't have impersonation information when first creating `MetastoreContext`. Ideally we should have a factory to handle this but it is outside of the scope of this PR.

**Review by shixuan-fan — APPROVED** [review]
LGTM
